src/genworld.cpp

```diff
@@ -326,3 +329,103 @@ void GenerateWorld(GenWorldMode mode, uint size_x, uint size_y, bool reset_setti
+	Json loaded = ParseJsonFromLoadedFile(detailed_err, TOWN_DATA_DIR);
+	/* If the parser failed, it'll set the detailed error which we'll pass on to the player. */
+	if (loaded.is_null()) {
+		SetDParamStr(0, detailed_err);
+		ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_JSON_PARSE_FAILED, WL_ERROR);
+		return;
+	}
+
+	/* Check for JSON formatting errors with the array of towns. */
+	if (!loaded.is_array()) {
+		ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_JSON_FORMATTED_INCORRECTLY, WL_ERROR);
+		return;
+	}
+
+	Json::array town_data = loaded.array_items();
+	uint failed_towns = 0;
+
+	/* Iterate through towns and attempt to found them. */
+	for (auto &feature : town_data) {
+		ExternalTownData town;
+
+		/* Check for errors in required JSON field "name".
+		 * Other fields are optional and if not present in the JSON file,
+		 * are set to safe values ("", 0, false) by the JSON '[type]_value()' function. */
+		if (!feature["name"].is_string()) {
+			ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_JSON_FORMATTED_INCORRECTLY, WL_ERROR);
+			return;
+		}
+
+		/* Set town properties. */
+		town.name = feature["name"].string_value();
+		town.population = feature["population"].int_value();
+		town.is_city = feature["iscity"].int_value() == 1;
+
+		/* Set town coordinates. */
+		town.x_percent = feature["x"].number_value();
+		town.y_percent = feature["y"].number_value();
+
+		/* If either coordinate is 0, the input was left blank or is not a number.
+		 * If either coordinate is greater than 1, the format is wrong. */
+		if (town.x_percent == 0 || town.y_percent == 0 || town.x_percent >= 1 || town.y_percent >= 1) {
+			SetDParamStr(0, town.name);
+			ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_BAD_COORDINATE, WL_ERROR);
+			return;
+		}
+
+		/* Find the target tile for the town. */
+		TileIndex tile = TileXY(town.x_percent * Map::MaxX(), town.y_percent * Map::MaxY());
+
+		/* Try founding on the target tile, and if that doesn't work, find the nearest suitable tile up to 16 tiles away.
+		 * The target might be on water, blocked somehow, or on a steep slope that can't be terraformed by the founding command. */
+		TileIndex search_tile = tile;
+		bool success = CircularTileSearch(&search_tile, 16, 0, 0, TryFoundTownNearby, &town);
+
+		/* If we still fail to found the town, we'll create a sign at the intended location and tell the player how many towns we failed to create in an error message. */
+		if (!success) {
+			Command<CMD_PLACE_SIGN>::Post(tile, town.name);
+			failed_towns++;
+			continue;
+		}
+
+		/* Now that we've created the town, let's grow it to the target population. */
+		Town* t = ClosestTownFromTile(search_tile, UINT_MAX);
+		uint tries = 500;
+
+		do {
+			Command<CMD_EXPAND_TOWN>::Post(t->index, 0);
+			if (tries-- == 0) break;
+		} while (t->cache.population < town.population);
```

Comment:
This is a bit weird .. only 500 tries? Why not 500 failures? Feels a bit random, 500.

Reply:
Why not 500 tries? This is to expand the town to the target population specified by the player in the JSON file, so we need some way to break out of the do/while loop in case their target population is not reachable. (Which is extremely likely, since some players will inevitably try to match real-world city populations of 1M+, which doesn't work in OpenTTD...)

Alternate suggestions welcome. 😃